var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -62,12 +82,78 @@ def install(self, spec, prefix):
                 g_args.append('--with-openmp')
             if 'qthreads' in spec:
                 g_args.append('--with-qthreads=%s' % spec['qthreads'].prefix)
             if 'cuda' in spec:
                 g_args.append('--with-cuda=%s' % spec['cuda'].prefix)
+            # host architectures
+            if 'host_arch=AMDAVX' in spec:
+                host_arch_args = 'AMDAVX'
+            if 'host_arch=ARMv80' in spec:
+                host_arch_args = 'ARMv80'
+            if 'host_arch=ARMv81' in spec:
+                host_arch_args = 'ARMv81'
+            if 'host_arch=ARMv8-ThunderX' in spec:
+                host_arch_args = 'ARMv8-ThunderX'
+            if 'host_arch=Power7' in spec:
+                host_arch_args = 'Power7'
+            if 'host_arch=Power8' in spec:
+                host_arch_args = 'Power8'
+            if 'host_arch=Power9' in spec:
+                host_arch_args = 'Power9'
+            if 'host_arch=WSM' in spec:
+                host_arch_args = 'WSM'
+            if 'host_arch=SNB' in spec:
+                host_arch_args = 'SNB'
+            if 'host_arch=HSW' in spec:
+                host_arch_args = 'HSW'
+            if 'host_arch=BDW' in spec:
+                host_arch_args = 'BDW'
+            if 'host_arch=SKX' in spec:
+                host_arch_args = 'SKX'
+            if 'host_arch=KNC' in spec:
+                host_arch_args = 'KNC'
+            if 'host_arch=KNL' in spec:
+                host_arch_args = 'KNL'
+            # gpu architectures
+            if 'gpu_arch=Kepler30' in spec:
+                gpu_arch_args = 'Kepler30'
+            if 'gpu_arch=Kepler32' in spec:
+                gpu_arch_args = 'Kepler32'
+            if 'gpu_arch=Kepler35' in spec:
+                gpu_arch_args = 'Kepler35'
+            if 'gpu_arch=Kepler37' in spec:
+                gpu_arch_args = 'Kepler37'
+            if 'gpu_arch=Maxwell50' in spec:
+                gpu_arch_args = 'Maxwell50'
+            if 'gpu_arch=Maxwell52' in spec:
+                gpu_arch_args = 'Maxwell52'
+            if 'gpu_arch=Maxwell53' in spec:
+                gpu_arch_args = 'Maxwell53'
+            if 'gpu_arch=Pascal60' in spec:
+                gpu_arch_args = 'Pascal60'
+            if 'gpu_arch=Pascal61' in spec:
+                gpu_arch_args = 'Pascal61'
+            # only a host architecture
+            if (host_arch_args!="" AND gpu_arch_args=""):
+                arch_args = '--arch='+host_arch_args
+            # only a gpu architecture
+            if (host_arch_args="" AND gpu_arch_args!=""):
+                if '+cuda' in spec:
+                    arch_args = '--arch='+gpu_arch_args
+            # both a host and a gpu architecture
+            if (host_arch_args!="" AND gpu_arch_args!=""):
+                if '+cuda' in spec:
+                    arch_args = '--arch='+host_arch_args+','+gpu_arch_args
+
+            if arch_args != "":
+                g_args += arch_args
```

Comment:
Will any of these ever be the empty string? It doesn't seem like that is a value these variants are allowed to take on. In that case, no need to do all of these checks.

Reply:
My only concern here is that the default for both types of arch args is 'none'. It is possible to build Kokkos without specifying an architecture. Specifying one is for "advanced" users that care about the implementation of atomics, having compiler vectorization flags specified, and such. I think that 'none' should stay as a sane default to protect less experienced users - as experts know to ask for what they want. The host/gpu arch args could be changed to check for !none, and this won't make the code less verbose, but it will make the intent more clear (and necessary with the suggested change to simplify the parsing above). Thanks again.